src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageSearch/PackageSearchRunner.cs

```diff
@@ -0,0 +1,201 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using NuGet.Common;
+using NuGet.Configuration;
+using NuGet.Protocol.Core.Types;
+using NuGet.Protocol;
+using NuGet.Commands;
+
+namespace NuGet.CommandLine.XPlat
+{
+    internal static class PackageSearchRunner
+    {
+        /// <summary>
+        /// Runs the search operation asynchronously using the provided parameters.
+        /// </summary>
+        /// <param name="sourceProvider">The provider for package sources.</param>
+        /// <param name="packageSearchArgs">Package search arguments</param>
+        /// <returns>A task that represents the asynchronous operation.</returns>
+        public static async Task RunAsync(
+            IPackageSourceProvider sourceProvider,
+            PackageSearchArgs packageSearchArgs,
+            CancellationToken cancellationToken)
+        {
+            var listEndpoints = GetPackageSourcesAsync(packageSearchArgs.Sources, sourceProvider);
+            WarnForHTTPSources(listEndpoints, packageSearchArgs.Logger);
+
+            Func<PackageSource, Task<IEnumerable<IPackageSearchMetadata>>> searchPackageSourceAsync =
+                packageSearchArgs.ExactMatch
+                ? packageSource => GetPackageAsync(packageSource, packageSearchArgs, cancellationToken)
+                : packageSource => SearchAsync(packageSource, packageSearchArgs, cancellationToken);
+
+            List<(Task<IEnumerable<IPackageSearchMetadata>>, PackageSource)> searchRequests = new(listEndpoints.Count);
+
+            foreach (var packageSource in listEndpoints)
+            {
+                Task<IEnumerable<IPackageSearchMetadata>> searchTask = searchPackageSourceAsync(packageSource);
+                searchRequests.Add((searchTask, packageSource));
+            }
+
+            IPackageSearchResultRenderer packageSearchResultRenderer;
+            packageSearchResultRenderer = new PackageSearchResultRendererTable(packageSearchArgs);
+            packageSearchResultRenderer.Start();
+
+            while (searchRequests.Count > 0)
+            {
+                Task<IEnumerable<IPackageSearchMetadata>> completedTask = await Task.WhenAny(searchRequests.Select(t => t.Item1));
+                int completedTaskIndex = searchRequests.FindIndex(t => t.Item1 == completedTask); ;
+                PackageSource source = searchRequests[completedTaskIndex].Item2;
+
+                await packageSearchResultRenderer.Add(source, completedTask);
+
+                searchRequests.RemoveAt(completedTaskIndex);
```

Comment:
nit, this is O(n^2) on list of sources as RemoveAt is O(n) as it has to copy the list. given you're never going to have more than a few sources, it doesn't matter, but otherwise should be a Hashset.